### PR TITLE
Add language extensions lab

### PR DESCRIPTION
**Pull request by pranjal-gururani**
## What does this PR do and why?

Add language extensions lab

## PR acceptance checklist

This checklist encourages us to confirm any changes have been analyzed to reduce risks in quality, performance, reliability, security, and maintainability.

- [X] Added/Updated documentation if applicable.
- [X] Pre-commit checks passed.
- [X] Lint and Nag checks passed.
- [X] If releasing a new version, have you bumped the version `make version part=<major|minor|patch>`?

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: content/intermediate/templates/language-extensions/index.md [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+2. From the left navigation panel, select the **Stacks** tab. Select the `language-extensions` stack you created earlier.
+3. From the top-right menu, choose **Update**.
+4. Under **Prerequisite - Prepare template,** select **Replace current template** and choose **Upload a template file**. Select **Choose file**, and supply the `language-extensions.yaml` template you updated earlier, and then choose **Next**.
+5. On **Specify Stack details** page, leave the configuration as it is. Choose **Next**.
+6. On **Configure Stack options**, leave the configuration as it is. Choose **Next**.
+7. On **Review** page, review the contents of the page. At the bottom of the page, choose to acknowledge all the capabilities shown in the **Capabilities and transforms** section.
+8. Choose **Submit**. Refresh the stack creation page until you see the stack in the `UPDATE_COMPLETE` status.
+9. Navigate to [CloudWatch console](https://console.aws.amazon.com/cloudwatch/). From the right navigation panel, choose **Dashboards**, **** and then **Dashboard** from the right navigation panel. You will see the dashboard you just created.
```
Comment (mrinaudo-aws) [inline content/intermediate/templates/language-extensions/index.md:162]:
```suggestion
9. Navigate to the [CloudWatch console](https://console.aws.amazon.com/cloudwatch/). From the right navigation panel, choose **Dashboards**. You should see the dashboard you just created.
```

File: content/intermediate/templates/language-extensions/index.md [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+
+
+Congratulations! You have learned how to use `Fn::ToJsonString`  to transform JSON objects into escaped JSON strings as inputs to resource properties.
```
Comment (mrinaudo-aws) [inline content/intermediate/templates/language-extensions/index.md:165]:
```suggestion
Congratulations! You have learned how to use `Fn::ToJsonString` to transform JSON objects into escaped JSON strings as inputs to resource properties.
```

File: content/intermediate/templates/language-extensions/index.md [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+
+### Challenge
+
+In this exercise, you’ll use the knowledge gained from earlier parts of this lab. Your task is to create an [Amazon Simple Storage Service (Amazon S3)](https://aws.amazon.com/s3/) bucket with its deletion policy set to a parameterized value of  `Delete`, and create a CloudWatch dashboard that reflects the number of objects in the bucket. Use the `language-extensions-challenge.yaml` template, and add content to it.
```
Comment (mrinaudo-aws) [inline content/intermediate/templates/language-extensions/index.md:169]:
```suggestion
In this exercise, you’ll use the knowledge gained from earlier parts of this lab. Your task is to create an [Amazon Simple Storage Service (Amazon S3)](https://aws.amazon.com/s3/) bucket with its deletion policy set to a parameterized value of `Delete`, and create a CloudWatch dashboard that reflects the number of objects in the bucket. Use the `language-extensions-challenge.yaml` template, and add content to it.
```

File: content/intermediate/templates/language-extensions/index.md [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+
+Refer to the [CloudWatch Dashboard structure](https://docs.aws.amazon.com/AmazonCloudWatch/latest/APIReference/CloudWatch-Dashboard-Body-Structure.html) mentioned in Part 2 of the lab as you describe the dashboard in your CloudFormation template. For the `metrics` field underneath `properties`, use `[[AWS/S3, NumberOfObjects, StorageType, AllStorageTypes, BucketName, !Ref S3Bucket]]` to denote the `NumberOfObjects` metrics in an S3 bucket for your CloudWatch widget. Please note that [S3 storage metrics are reported once per day](https://docs.aws.amazon.com/AmazonS3/latest/userguide/cloudwatch-monitoring.html) at no additional cost, so you may not see them when you are running the lab.
+
+
+:::expand{header="Need a hint?"}
+* Recall using the language extension in Part 1 of the lab to use a parameter for the deletion policy.
+* Don’t forget to reference the deletion policy parameter in your S3 bucket resource.
+* Additionally, recall how you added a CloudWatch dashboard earlier, add use the `NumberOfObjects`  metrics for the relevant field.
```
Comment (mrinaudo-aws) [inline content/intermediate/templates/language-extensions/index.md:177]:
```suggestion
* Additionally, recall how you added a CloudWatch dashboard earlier, add use the `NumberOfObjects` metrics for the relevant field.
```

File: code/workspace/language-extensions/language-extensions-challenge.yaml [hunk 5]
```diff
@@ -0,0 +1,16 @@
+AWSTemplateFormatVersion: "2010-09-09"
+Transform: AWS::LanguageExtensions
```
Comment (mrinaudo-aws) [inline code/workspace/language-extensions/language-extensions-challenge.yaml:2]:
```suggestion
```

File: content/intermediate/templates/language-extensions/index.md [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+:::
+
+:::expand{header="Want to see the solution?"}
```
Comment (mrinaudo-aws) [inline content/intermediate/templates/language-extensions/index.md:180]:
```suggestion
:::expand{header="Want to see the solution?"}
* Add the `Transform: AWS::LanguageExtensions` line to the template like you did in Part 1 of the lab.
```

File: content/intermediate/templates/language-extensions/index.md [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+* Edit the `Parameters` section to add the `DeletionPolicyParameter` like you did in Part 1 of the lab.
+* Underneath the `Resources` section for the `S3Bucket` resource, add the `DeletionPolicy` attribute with a reference to the parameter.
+* Underneath the `Resources` section, add the `Dashboard` resource.
+* You can find the full language-extensions-solution in the template `language-extensions-solution.yaml` in the `code/solutions/language-extensions` directory.
```
Comment (mrinaudo-aws) [inline content/intermediate/templates/language-extensions/index.md:184]:
```suggestion
* You can find the full challenge solution in the template called `language-extensions-solution.yaml`, that is in the `code/solutions/language-extensions` directory.
```

File: content/intermediate/templates/language-extensions/index.md [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+
+1. Navigate to the [AWS CloudFormation Console](https://console.aws.amazon.com/cloudformation/).
+2. On the **Stacks** page in the CloudFormation console, select the stack you created in **Part 1:** `language-extensions`.
+3. In the stack details pane, choose **Delete**.  Select **Delete stack** when prompted.
+4. On the **Stacks** page in the CloudFormation console, select the stack you created in **Challenge** section: `language-extensions-solution`.
+5. In the stack details pane, choose **Delete**.  Select **Delete stack** when prompted.
+
+### Conclusion
+
+Great work! You learned how to incorporate `AWS::LanguageExtensions` in your CloudFormation template. Please feel free to leave us your feedback at our [Language Discussion GitHub repository](https://github.com/aws-cloudformation/cfn-language-discussion). We welcome your contributions to RFCs!
```
Comment (mrinaudo-aws) [inline content/intermediate/templates/language-extensions/index.md:200]:
```suggestion
Great work! You learned how to incorporate `AWS::LanguageExtensions` in your CloudFormation templates. Please feel free to provide feedback for RFCs in the [Language Discussion GitHub repository](https://github.com/aws-cloudformation/cfn-language-discussion). We welcome your contributions!
```

**Review by mrinaudo-aws — CHANGES_REQUESTED** [review]
Thank you, @pranjal-gururani !  Please review/accept the suggestions, and ask for another review!

**Review by mrinaudo-aws — CHANGES_REQUESTED** [review]
Thank you, @pranjal-gururani !  I added minor suggestions for you to review/apply; please request another review when done!

File: code/solutions/language-extensions/language-extensions-part2.yaml [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+    Type: AWS::CloudWatch::Dashboard
+    Properties:
+      DashboardBody:
+        Fn::ToJsonString:
+          start: -PT6H
+          periodOverride: inherit
+          widgets:
+            - type: metric
+              x: 0
+              "y": 7
```
Comment (mrinaudo-aws) [inline code/solutions/language-extensions/language-extensions-part2.yaml:35]: @pranjal-gururani should this not have the quotes, for consistency?

File: content/intermediate/templates/language-extensions/index.md [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+    Type: AWS::CloudWatch::Dashboard
+    Properties:
+      DashboardBody:
+        Fn::ToJsonString:
+          start: -PT6H
+          periodOverride: inherit
+          widgets:
+            - type: metric
+              x: 0
+              "y": 7
```
Comment (mrinaudo-aws) [inline content/intermediate/templates/language-extensions/index.md:134]: Remove quotes.

File: code/solutions/language-extensions/language-extensions-part2.yaml (hunk 9 above)
Comment (mrinaudo-aws) [reply code/solutions/language-extensions/language-extensions-part2.yaml:35]: Actually - my bad, the formatter we use would have added the quotes, and I believe this is expected -and makes sense- to distinguish a string -like in your case- from a value that might be interpreted as a boolean.  I think you can leave it as it is or, for consistency, you could try adding quotes in other parts as well in the widget (unless the formatter expunges the quotes, deeming them not needed).  Let me know what you think!

File: code/solutions/language-extensions/language-extensions-solution.yaml [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+    Type: AWS::CloudWatch::Dashboard
+    Properties:
+      DashboardBody:
+        Fn::ToJsonString:
+          start: -PT6H
+          periodOverride: inherit
+          widgets:
+            - type: metric
+              x: 0
+              "y": 7
```
[the remark this answers is not in the document]
Comment (mrinaudo-aws) [reply code/solutions/language-extensions/language-extensions-solution.yaml:37]: My bad - see my comment: https://github.com/aws-samples/cfn101-workshop/pull/229/files#r1192530058

File: content/intermediate/templates/language-extensions/index.md (hunk 10 above)
Comment (mrinaudo-aws) [reply content/intermediate/templates/language-extensions/index.md:134]: Sorry, my bad - see my comment: https://github.com/aws-samples/cfn101-workshop/pull/229/files#r1192530058

File: code/solutions/language-extensions/language-extensions-part2.yaml (hunk 9 above)
Comment (pranjal-gururani) [reply code/solutions/language-extensions/language-extensions-part2.yaml:35]: I am going to leave "y" quoted, as the formatter will add the other quotes.

**Review by mrinaudo-aws — DISMISSED** [review]
LGTM

**Comment by mrinaudo-aws** [comment]
> Hey folks, please resolve 8 unresolved conversations to be able to merge.
> 
> FYI, after we merge this, the release will come little bit later as we depend on #224 and #225 to cover our Full Day agenda. Apologies for the blocker.

Thank you, @rezabekf - most of them were stale.  I resolved them.

**Review by rezabekf — APPROVED** [review]
Thank you. Approved. This will be released in the next release window with other PRs. ETA: this week

**Pull request merged**